### PR TITLE
dealii: add nanoflann, sundials and adol-c  plus fixes on macOS

### PR DESCRIPTION
**Pull request by davydden**
fixes https://github.com/LLNL/spack/issues/5173

p.s. @markcmiller86 in 39cb589510e95fa44ca634b2cdac9a56a34b1809 you extended `sundials` a bit and it looks like it worked for you on macOS. Could you please check if this PR does not break it for you?

### CODE REVIEW

File: var/spack/repos/builtin/packages/sundials/package.py [hunk 1]
```diff
@@ -68,10 +68,18 @@ def on_off(varstr):
             return 'ON' if varstr in self.spec else 'OFF'
 
         cmake_args = std_cmake_args[:]
+
+        fortran_flag = '{0}'.format(self.compiler.pic_flag)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sundials/package.py]: You don't need to convert this to a string, it already is a string.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -109,6 +115,7 @@ class Dealii(CMakePackage):
 
     # optional dependencies
     depends_on("mpi",              when="+mpi")
+    depends_on("adol-c@2.6.4:",    when='@9.0:+adol-c')
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: They have not yet made a point release that fixes a really minor issue that affects print functions / our test suite. We either need to require `adol-c@dev` for now or we could just apply [this little patch](https://gitlab.com/adol-c/adol-c/commit/9af81273bfa01ea37e6a4e1893dd5e07fea6e467) to `adol-c@2.6.3`.

Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: i say it's fine as is, we have not released `9.0` either, right? :smile:

Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py]: Agreed!

**Comment by davydden** [comment]
@adamjstewart can we merge this?

File: var/spack/repos/builtin/packages/expat/package.py [hunk 3]
```diff
@@ -39,7 +40,7 @@ class Expat(AutotoolsPackage):
     # someone's asking for an older version and also libbsd.
     # In order to install an older version, you'll need to add
     # `~libbsd`.
-    variant('libbsd', default=True,
+    variant('libbsd', default=sys.platform != 'darwin',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/expat/package.py]: Is this change related to this PR?

File: var/spack/repos/builtin/packages/sundials/package.py [hunk 4]
```diff
@@ -68,10 +68,18 @@ def on_off(varstr):
             return 'ON' if varstr in self.spec else 'OFF'
 
         cmake_args = std_cmake_args[:]
+
+        fortran_flag = self.compiler.pic_flag
+        if spec.satisfies('%clang') and (sys.platform == 'darwin'):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sundials/package.py]: You could simplify this as `spec.satisfies('%clang platform=darwin')`.

File: var/spack/repos/builtin/packages/sundials/package.py [hunk 5]
```diff
@@ -68,10 +68,18 @@ def on_off(varstr):
...
+            libgfortran = os.path.dirname(os.popen(
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sundials/package.py]: According to the [documentation](https://docs.python.org/2/library/os.html#os.popen), `os.popen` has been deprecated since Python 2.6. It would be better to use Spack's `Executable` class. Something like:
```python
mpif77 = Executable(self.spec['mpi'].mpif77)
libgfortran = LibraryList(mpif77('--print-file-name', 'libgfortran.a', output=str))
fortran_flag += ' ' + libgfortran.ld_flags
```

File: var/spack/repos/builtin/packages/expat/package.py (hunk 3 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/expat/package.py]: yes, otherwise i can't install doxygen and consequently dealii

**Comment by davydden** [comment]
@adamjstewart let me see if I did not break anything while doing cleanup...

**Comment by davydden** [comment]
@adamjstewart yes, `sundials` still works fine on Sierra with clang+gfortran.

**Comment by davydden** [comment]
p.s. you can keep commits separate as they are self contained

**Comment by markcmiller86** [comment]
Apologies...I did not have time to investigate on my mac last night. But, I see @adamjstewart did so am fine with that confirmation.

**Comment by davydden** [comment]
@adamjstewart ping.
